### ISSUE TITLE
TaskInputFilePropertyBuilder should allow setting @[Runtime]Classpath

### ISSUE THREAD

**Issue by ducrohet** [gradle/gradle#2968]
In the Android plugin we have a task that handle inputs dynamically. This is our Transform task that gets fed inputs from a pipeline so the number of inputs depends on the transform itself, and the task is generic.

We use to merge all these inputs into a single FileCollection and returning this in an `@InputFiles` annotated getter but this doesn't work when enabling the task for caching.

For caching support, we are putting `PathSensitivity.RELATIVE` on the input but this changes the behavior and content of `IncrementalTaskInputs` that our task receives. For example, if a file moved from one input folder to another, we are not receiving not receiving a removed+added anymore. Basically all the input folders are considered a single input.

We needed to fix this so we are now looping on all the input streams of the transform and doing this:
`getInputs().files(stream.getAsFileTree()).withPathSensitivity(PathSensitivity.RELATIVE);`

This works great, up to a point.

We are now wanting to fix another issue related to caching, which is that some of the inputs to the transforms are Jar files, and we want to ignore things like timestamp, so we want to use `@RuntimeClasspath`. However the `TaskInputFilePropertyBuilder` does not allow us to set this.

Until this is possible our transforms will not be cacheable. Or rather they are declared as cacheable but because upstream jar-based tasks are not cacheable and will re-run again, even in a 100% cache hit build, they will run again too because their inputs will not match the cache key.

Could this fix be fixed?

**Comment by oehme** [comment]
I think the runtime API should have all the same capabilities that the annotation based API has. @lptr is that your opinion as well?

**Comment by wolfs** [comment]
@ducrohet You already can do `getInputs().files(stream.getAsFileTree()).withSnapshotter(ClasspathSnapshotter.class)` by calling `TaskFilePropertyBuilderInternal.withSnapshotter`. That is an internal API. We should add these capabilities to the public API as well.

**Comment by ducrohet** [comment]
Is that internal API fairly stable? We're ok to use it for a little bit if we know that a public API is coming soon (4.3/4.4)

**Comment by lptr** [comment]
I don't think we want to open up the `withSnapshotter()` API. It doesn't make sense to open it up without allowing users to declare their own snapshotters, and that would be way too complex. We have better ways to deal with this I think.

We have a goal in Gradle 4.3 to bring the annotation API (`@InputFile`) and the runtime API (`task.inputs.file(...)`) closer together. Adding runtime API capability for registering `@Classpath` properties seems to fall into that bucket, too. However, we'd like to do this in an extensible way without making a ton of internal APIs public.

We are going to implement what we call "polymorphic input support". Currently, the `@Nested` annotation only supports discovering nested properties based on the static type of the `@Nested` property itself. This will change in 4.3, and we'll support discovering nested properties based on the runtime type of property values. This means that you'll be able to do something like this:

```groovy
@Nested
Map<String, Object> getComplexInputs() {
    return [
        classes: new ClasspathInput(...)
    ]
}

class ClasspathInput {
    @Classpath
    FileCollection classpath
}
```

...and Gradle will find the nested classpath despite the fact that the `ClasspathInput` type is not declared on the task level. We might even end up having something like `task.inptus.nested(...)` eventually.

Would this work for you to declare your classpaths?

**Comment by ducrohet** [comment]
Would those be considered different inputs in this case? ie if `getComplexInputs()` is annotated with `@PathSensitivity.RELATIVE` and a file from one input move into another input with the same relative path, will it be seen as a `REMOVED` and `ADDED`?

If that's the case then yes it would work for us. There's a bit more layers of encapsulated objects than we'd like for this but this is fine.

I do like the change to the behavior of `@Nested` using the runtime type too.

**Comment by oehme** [comment]
@lptr Why does `withSnapshotter` only make sense if it is user-extensible? It feels to me like a reasonable addition even if the set of annotations you can pass there is fixed for now. Also long term we should consider opening up the snapshotter API to allow languages like Kotlin to do their own compile avoidance. So why not take a first step in that direction?

**Comment by lptr** [comment]
We've discussed this with @adammurdoch and @wolfs yesterday, and agreed to implement a simpler solution first. This will most probably be opening up `withSnapshotter()`, but we won't be opening up any APIs to implement your own snapshotters at this point.

**Comment by lptr** [comment]
PR has been merged, this will ship in Gradle 4.3.
